File: grant_clean_js/68.js
const cheerio = require('cheerio');

function getTitle(cheer) {
    try {
        if(cheer("h1.readcontent").first().text().trim() == "")
            return null;
        return cheer("h1.readcontent").first().text().trim();
        // return soup.find("h1", {"class" { "readcontent"}).get_text().trim()
    }
    catch {
        return null;
    }
}

function getProjectId(cheer) {
    var candSection = [];
    try {
        candSection = cheer("div.section.emphasize").children();
        // .each(function(i, elem){
        //     candSection.push(cheer(this));
        // });
        
        // candSection = soup.find("div", {"class" { "section emphasize"}).find_all()
    }
    catch {
        return null;
    }
    var projectId = null;
    // for (i in range(len(candSection) - 1)) {
     for(let i=0; i<candSection.length-1; i++){
        var curTag  = candSection[i]
        if (i + 1 < candSection.length) {
            var nextTag = candSection[i + 1];
        }
        else {
            var nextTag = null;
        }
        if (curTag.children[0].data.toString().trim() == "Project number" && nextTag != null){
            // if (curTag.name.startswith("h") && nextTag.name.startswith("p")) {
            if(curTag.name.startsWith("h") && nextTag.name.startsWith("p")) {
                projectId = nextTag.children[0].data.toString().trim();
            }
        }
    
    }
    return projectId;
}

function getResearcher(cheer) {
    var candSection = [];
    try {
        candSection = cheer("div.section.emphasize").children();
        // .each(function(i, elem){
        //     candSection.push(cheer(this));
        // });
        // candSection = soup.find("div", {"class" { "section emphasize"}).find_all()
    }
    catch {
        return null;
    }
    var researcher = [];
    for (let i=0; i < candSection.length - 1; i++) {
        var curTag  = candSection[i];
        if (i + 1 < candSection.length) {
            var nextTag = candSection[i + 1];
        }
        else {
            nextTag = null;
        }
        if (curTag.children[0].data.toString().trim() == "Main applicant" && nextTag != null){
            if (curTag.name.startsWith("h") && nextTag.name.startsWith("p")) {
                researcher.push(nextTag.children[0].data.toString().trim());
            }
        }
        if (curTag.children[0].data.toString().trim() == "Team members" && nextTag != null){
            if (curTag.name.startsWith("h") && nextTag.name.startsWith("p")) {
                for( res of nextTag.children[0].data.toString().trim().split(",")){
                    researcher.push(res.trim());
                }
            }
        }
    }
    if (researcher.length == 0) {
        return null;
    }
    return researcher.join(";");
    // return ";".join(researcher);
}

function getInstitution(cheer) {
    var candSection = [];
    try {
        candSection = cheer("div.section.emphasize").children();
        // .each(function(i, elem){
        //     candSection.push(cheer(this));
        // });
        // candSection = soup.find("div", {"class" { "section emphasize"}).find_all()
    }
    catch {
        return null;
    }
    var institution = null;
    for (let i=0; i<candSection.length - 1; i++){
        var curTag  = candSection[i];
        if (i + 1 < candSection.length ){
            var nextTag = candSection[i + 1];
        }
        else {
            var nextTag = null;
        }
        if (curTag.children[0].data.toString().trim() == "Affiliated with" && nextTag != null){
            if (curTag.name.startsWith("h") && nextTag.name.startsWith("p")) {
                institution = nextTag.children[0].data.toString().trim();
            }
        }
    }
     
    if (institution == "") {
        return null;
    }

    return institution;
}
function getDescription(cheer) {
    var description = [];
    try {
        cheer("div.readcontent").find("p").each(function(i, elem){
            description.push(cheer(this).text());
        });
        // description_cand = soup.find("div", {"class" { "readcontent"})
    }
    catch {
        return null;
    }
    // var description = [];

    // try {
    //     description = description_cand.find_all("p")
    // catch {
    //     pass

    if (description.length > 0) {
        let tmpDesc = [];
        for(desc of description){
            tmpDesc.push(desc.toString().trim());
        }
        description = tmpDesc.join("\n").trim();
        //description = "\n".join([desc.get_text().trim() for desc in description]).trim()
    }
    else {
        return null;
    }
    if (description == "") {
        return null;
    }
    return description;
}
function getDescriptionRaw(cheer) {
    var descriptionRaw = null;
    try {
        descriptionRaw = cheer("div.readcontent").first().text().trim();
        // description_raw_cand = soup.find("div", {"class" { "readcontent"})
    }
    catch {
        return null;
    }

    if (descriptionRaw == "") {
        return null;
    }
    return descriptionRaw;
}

function getStartEndDate(cheer) {
    var candSection = [];
    try {
        candSection = cheer("div.section.emphasize").children();
        // .each(function(i, elem){
        //     candSection.push(cheer(this));
        // });
        // // candSection = soup.find("div", {"class" { "section emphasize"}).find_all()
    }
    catch {
        return [null, null];
    }
    
    var startEndDate = ""
    for (let i=0; i < candSection.length - 1;i++) {
        var curTag  = candSection[i];
        if (i + 1 < candSection.length ){
         var nextTag = candSection[i + 1];
        }
        else {
            var nextTag = null;
        }
        if (curTag.children[0].data.toString().trim() == "Duration" && nextTag != null){
            if (curTag.name.startsWith("h") && nextTag.name.startsWith("p")) {
                startEndDate = nextTag.children[0].data.toString().trim();
            }
        }
    }
    
    var regexString = /\b\d{2}\/\d{2}\/\d{4}\b/g;
    var matchs = [];
    matchs = startEndDate.match(regexString);
    // matchs = re.findall(regexString, startEndDate)
    if(matchs != null)
    if (matchs.length > 0) {
        for (let i=0; i<matchs.length;i++) {
            try {
                var dateSplitted = matchs[i].split("/");
                var toDate = new Date(dateSplitted[2],(parseInt(dateSplitted[1])-1).toString(),(parseInt(dateSplitted[0])+1).toString());
                // matchs[i] = datetime.datetime.strptime(matchs[i], "%d/%m/%Y"):
                matchs[i] = toDate.toISOString().substring(0,10).replace(/-/g,"/");
            }
            catch {
                matchs[i] = null;
            }
        }
    }
    else {
        return [null, null];
    }

    var startDate = null;
    var endDate = null;
    if(matchs != null)
    if (matchs.length == 1) {
        startDate = matchs[0];
    }
    else if(matchs.length > 1) {
        startDate = matchs[0];
        endDate = matchs[1];
    }

    return [startDate, endDate];
}

function getStartYear(startDate) {
    if (startDate == null) 
        return null;
    else if(startDate.length == 0 )
        return null;
    return startDate.substr(0,4)
    // return startDate[ : 4];
    
}
function getEndYear(endDate) {
    if (endDate.length == 0 || endDate == null) 
        return null;
    return endDate.substr(0,4);
    // return endDate[ : 4]
}


function cleanHtml(html){
    return html.toString().replace(/\n|\t|\s+/g," ");
    //re.sub(r"\n|\t|\s+", " ", str(soup))
}

function parse(features) { 
    var html = features['html'];
    if(html == null){
        return features;
    }
    html = cleanHtml(html);
    var cheer = cheerio.load(html, {decodeEntities: false});
    
    features["title"]               = getTitle(cheer);
    features["project_id"]          = getProjectId(cheer);
    features["project_id_original"] = features["project_id"];

    features["researcher"]      = getResearcher(cheer);
    features["researcher_en"]   = features["researcher"];
    features["institution"]     = getInstitution(cheer);

    features["description_en"]  = getDescription(cheer);
    features["description"]     = features["description_en"];
    features["description_raw"] = getDescriptionRaw(cheer);
    var dates = getStartEndDate(cheer);
    features["start_date"] = dates[0];
    features["end_date"] = dates[1];
    features["start_year"] = getStartYear(features["start_date"]);
    features["end_year"]   = getStartYear(features["end_date"]);
    features["country"] = "NL";

    if( features["start_year"] != null && features["end_year"] !=null ) {
        features["project_term"] = parseInt(features["end_year"]) - parseInt(features["start_year"]) + 1 + " years"
    }else{
        features["project_term"] = null
    };
    

    return features

}

module.exports = {
    parse,
  };

File: grant_clean/68.py
# cnki parser
import re
import os
import csv
import time
import datetime
import configparser
from tools import read_tsv
from bs4 import BeautifulSoup
from multiprocessing import Pool, cpu_count

from db import sql_execute, sql_insert, sql_insert_many
from tools import write_to_csv

def clean_row(row) :
    features = {
        "id"                : row["id"],
        "grant_sites_id"    : row["grant_sites_id"],
        "project_id"        : "NOT_FOUND",
        "project_id_original" : "NOT_FOUND",
        "title"             : "NOT_FOUND",
        "title_en"          : "NOT_FOUND",
        "project_url"       : row["project_url"],
        "keyword"           : "NOT_FOUND",
        "keyword_en"        : "NOT_FOUND",
        "source"            : "NOT_FOUND",
        "researcher"        : "NOT_FOUND",
        "researcher_en"     : "NOT_FOUND",
        "institution"       : "NOT_FOUND",
        "country"           : "NL",
        "currency"          : "NOT_FOUND",
        "award_amount"      : None,
        "award_amount_usd"  : None,
        "funded_date"       : None,
        "start_year"        : "NOT_FOUND",
        "end_year"          : "NOT_FOUND",
        "start_date"        : None,
        "end_date"          : None,
        "project_term"      : "NOT_FOUND",
        "description"       : "NOT_FOUND",
        "description_raw"   : "NOT_FOUND",
        "description_en"    : "NOT_FOUND",    
    }
    html = row["html"]
    del row

    if html == "NOT_FOUND" :
        # print("html not found")
        return features
    
    try :
        soup = BeautifulSoup(html, "html.parser")
    except :
        # print("Soup errors!")
        return features
    
    features["title"]               = get_title(soup)
    features["project_id"]          = get_project_id(soup)
    features["project_id_original"] = features["project_id"]

    features["researcher"]      = get_researcher(soup)
    features["researcher_en"]   = features["researcher"]
    features["institution"]     = get_institution(soup)

    features["description_en"]  = get_description(soup)
    features["description"]     = features["description_en"]
    features["description_raw"] = get_description_raw(soup)

    features["start_date"], features["end_date"] = get_start_end_date(soup)
    features["start_year"] = get_start_year(features["start_date"])
    features["end_year"]   = get_start_year(features["end_date"])

    return features

def get_title(soup) :
    try :
        return soup.find("h1", {"class" : "readcontent"}).get_text().strip()
    except :
        return "NOT_FOUND"
    return "NOT_FOUND"

def get_project_id(soup) :
    try :
        cand_section = soup.find("div", {"class" : "section emphasize"}).find_all()
    except :
        return "NOT_FOUND"
    
    project_id = "NOT_FOUND"
    for i in range(len(cand_section) - 1) :
        cur_tag  = cand_section[i]
        if i + 1 < len(cand_section) :
            next_tag = cand_section[i + 1]
        else :
            next_tag = None
        if cur_tag.get_text().strip() == "Project number" and next_tag is not None:
            if cur_tag.name.startswith("h") and next_tag.name.startswith("p") :
                project_id = next_tag.get_text().strip()

    return project_id

def get_researcher(soup) :
    try :
        cand_section = soup.find("div", {"class" : "section emphasize"}).find_all()
    except :
        return "NOT_FOUND"
    
    researcher = []
    for i in range(len(cand_section) - 1) :
        cur_tag  = cand_section[i]
        if i + 1 < len(cand_section) :
            next_tag = cand_section[i + 1]
        else :
            next_tag = None
        if cur_tag.get_text().strip() == "Main applicant" and next_tag is not None:
            if cur_tag.name.startswith("h") and next_tag.name.startswith("p") :
                researcher.append(next_tag.get_text().strip())
        if cur_tag.get_text().strip() == "Team members" and next_tag is not None:
            if cur_tag.name.startswith("h") and next_tag.name.startswith("p") :
                for res in next_tag.get_text().strip().split(",") :
                    researcher.append(res.strip())
    if not researcher :
        return "NOT_FOUND"
    return ";".join(researcher)

def get_institution(soup) :
    try :
        cand_section = soup.find("div", {"class" : "section emphasize"}).find_all()
    except :
        return "NOT_FOUND"
    
    institution = "NOT_FOUND"
    for i in range(len(cand_section) - 1) :
        cur_tag  = cand_section[i]
        if i + 1 < len(cand_section) :
            next_tag = cand_section[i + 1]
        else :
            next_tag = None
        if cur_tag.get_text().strip() == "Affiliated with" and next_tag is not None:
            if cur_tag.name.startswith("h") and next_tag.name.startswith("p") :
                institution = next_tag.get_text().strip()
     
    if institution == "" :
        return "NOT_FOUND"

    return institution

def get_description(soup) :
    try :
        description_cand = soup.find("div", {"class" : "readcontent"})
    except :
        return "NOT_FOUND"
    description = []

    try :
        description = description_cand.find_all("p")
    except :
        pass

    if description :
        description = "\n".join([desc.get_text().strip() for desc in description]).strip()
    else :
        return "NOT_FOUND"
    if description == "" :
        return "NOT_FOUND"
    return description

def get_description_raw(soup) :
    try :
        description_raw_cand = soup.find("div", {"class" : "readcontent"})
    except :
        return "NOT_FOUND"
    description_raw = "NOT_FOUND"

    try :
        description_raw = description_raw_cand.get_text().strip()
    except :
        pass

    if description_raw == "" :
        return "NOT_FOUND"
    return description_raw

def get_start_end_date(soup) :
    try :
        cand_section = soup.find("div", {"class" : "section emphasize"}).find_all()
    except :
        return None, None
    
    start_end_date = ""
    for i in range(len(cand_section) - 1) :
        cur_tag  = cand_section[i]
        if i + 1 < len(cand_section) :
            next_tag = cand_section[i + 1]
        else :
            next_tag = None
        if cur_tag.get_text().strip() == "Duration" and next_tag is not None:
            if cur_tag.name.startswith("h") and next_tag.name.startswith("p") :
                start_end_date = next_tag.get_text().strip()
    
    regex_string = r"\b\d{2}/\d{2}/\d{4}\b"
    matchs = re.findall(regex_string, start_end_date)
    if matchs :
        for i in range(len(matchs)) :
            try :
                matchs[i] = datetime.datetime.strptime(matchs[i], "%d/%m/%Y")
            except :
                matchs[i] = None
                pass
    else :
        return None, None

    start_date, end_date = None, None
    # print(matchs)
    if len(matchs) == 1 :
        start_date = str(matchs[0])
    elif len(matchs) > 1 :
        start_date, end_date = str(matchs[0]), str(matchs[1])
    # print(start_date, end_date)
    return start_date, end_date

def get_start_year(start_date) :
    if start_date is None :
        return "NOT_FOUND"
    return start_date[ : 4]

def get_end_year(end_date) :
    if end_date is None :
        return "NOT_FOUND"
    return end_date[ : 4]

def multi_process_db(grant_sites_id) :
    start_time = time.time()

    sql = """SELECT id, grant_sites_id, project_url FROM tmp_other_info WHERE grant_sites_id = %s"""
    query_result = sql_execute(sql, db_config, (grant_sites_id, ))

    hash_url = {}

    for row in query_result :
        hash_url[str(row[2]).strip()] = 1

    sub_grant_datas = []
    
    sql = """SELECT id, grant_sites_id, project_url, html FROM tmp_other_html WHERE grant_sites_id = %s"""
    query_result = sql_execute(sql, db_config, (grant_sites_id, ))

    for row in query_result :
        features = {
            "id"              : row[0],
            "grant_sites_id"  : row[1],
            "project_url"     : str(row[2]).strip(),
            "html"            : row[3],
        }
        if features["project_url"] not in hash_url :
            sub_grant_datas.append(features)
            hash_url[features["project_url"]] = 1

    p = Pool(processes=cpu_count())
    grant_datas = p.map(clean_row, sub_grant_datas)
    p.close()
    p.join()

    sql = """INSERT INTO tmp_other_info(
        id,
        grant_sites_id,
        project_id,
        project_id_original,
        title,
        title_en,
        project_url,
        keyword,
        keyword_en,
        source,
        researcher,
        researcher_en,
        institution,
        country,
        currency,
        award_amount,
        award_amount_usd,
        funded_date,
        start_year,
        end_year,
        start_date,
        end_date,
        project_term,
        description,
        description_raw,
        description_en,
        created_date,
        updated_date,
        crawled_date)
        VALUES(%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s, now(),now(),now() )"""
    
    input_datas = []
    
    for row in grant_datas :
        input_datas.append((
            int(row["id"]),
            int(row["grant_sites_id"]),
            row["project_id"],
            row["project_id_original"],
            row["title"],
            row["title_en"],
            row["project_url"],
            row["keyword"],
            row["keyword_en"],
            row["source"],
            row["researcher"],
            row["researcher_en"],
            row["institution"],
            row["country"],
            row["currency"],
            row["award_amount"],
            row["award_amount_usd"],
            row["funded_date"],
            row["start_year"],
            row["end_year"],
            row["start_date"],
            row["end_date"],
            row["project_term"],
            row["description"],
            row["description_raw"],
            row["description_en"]
        ))

    sql_insert_many(sql, db_config, input_datas)
    print(str(time.time() - start_time))

if __name__ == "__main__":
    start_time = time.time()

    config = configparser.ConfigParser()
    config.read('.env')

    db_config = (config["DEFAULT"]["_HOSTNAME"], config["DEFAULT"]["_DATABASE"], config["DEFAULT"]["_USERNAME"], \
    config["DEFAULT"]["_PASSWORD"], config["DEFAULT"]["_PORT"])

    multi_process_db(68)

    print("Total time is : ", str(time.time() - start_time))